evap/staff/importers/json.py

```diff
@@ -0,0 +1,325 @@
+import json
+import logging
+from dataclasses import dataclass, field
+from datetime import datetime, timedelta
+from typing import TypedDict
+
+from django.db import transaction
+from django.utils.timezone import now
+
+from evap.evaluation.models import Contribution, Course, CourseType, Degree, Evaluation, Semester, UserProfile
+from evap.evaluation.tools import clean_email
+from evap.staff.tools import update_or_create_with_changes, update_with_changes
+
+logger = logging.getLogger("import")
+
+
+class ImportStudent(TypedDict):
+    gguid: str
+    email: str
+    name: str
+    christianname: str
+
+
+class ImportLecturer(TypedDict):
+    gguid: str
+    email: str
+    name: str
+    christianname: str
+    titlefront: str
+
+
+class ImportCourse(TypedDict):
+    cprid: str
+    scale: str
+
+
+class ImportRelated(TypedDict):
+    gguid: str
+
+
+class ImportAppointment(TypedDict):
+    begin: str
+    end: str
+
+
+class ImportEvent(TypedDict):
+    gguid: str
+    lvnr: int
+    title: str
+    title_en: str
+    type: str
+    isexam: bool
+    courses: list[ImportCourse]
+    relatedevents: ImportRelated
+    appointments: list[ImportAppointment]
+    lecturers: list[ImportRelated]
+    students: list[ImportRelated]
+
+
+class ImportDict(TypedDict):
+    students: list[ImportStudent]
+    lecturers: list[ImportLecturer]
+    events: list[ImportEvent]
+
+
+@dataclass
+class NameChange:
+    old_last_name: str
+    old_first_name_given: str
+    new_last_name: str
+    new_first_name_given: str
+
+
+@dataclass
+class ImportStatistics:
+    name_changes: list[NameChange] = field(default_factory=list)
+    new_courses: list[Course] = field(default_factory=list)
+    new_evaluations: list[Evaluation] = field(default_factory=list)
+    updated_courses: list[Course] = field(default_factory=list)
+    updated_evaluations: list[Evaluation] = field(default_factory=list)
+    attempted_changes: list[Evaluation] = field(default_factory=list)
+
+    @staticmethod
+    def _make_heading(heading: str) -> str:
+        heading += "\n" + "".join(["-" for i in heading]) + "\n"
+        return heading
+
+    @staticmethod
+    def _make_total(total: int) -> str:
+        return f"({total} in total)\n\n"
+
+    def get_log(self) -> str:
+        log = "JSON IMPORTER REPORT\n"
+        log += "====================\n\n"
+        log += f"Import finished at {now()}\n\n"
+        log += self._make_heading("Name Changes")
+        for name_change in self.name_changes:
+            log += f"- {name_change.old_first_name_given} {name_change.old_last_name} → {name_change.new_first_name_given} {name_change.new_last_name}\n"
+        log += self._make_total(len(self.name_changes))
```

Comment:
For readability, I'd put the count close towards the heading. If its hundreds of changes here, it's hard to associate the number.